File: scripts/generate-og-image.py
#!/usr/bin/env python3
"""Generate public/og-image.png (1200x630).

The companion generate-og-image.js is the canonical description of this card
and renders it through a real browser. It needs Node and Playwright, neither of
which exists on the machine this repo is maintained from, so this script draws
the same layout with Pillow instead. It is what produced the committed PNG.

Keep the two in step: if the design changes, change both.
"""
from PIL import Image, ImageDraw, ImageFont
import os

# The card is laid out in 1200x630 design units and rendered at SCALE times
# that. Link previews are displayed on high-DPI screens at roughly twice their
# CSS size, so a 1x asset gets resampled and text goes soft; 2x lands sharp.
# Every measurement below stays in design units and is multiplied on the way out.
SCALE = 2
W, H = 1200 * SCALE, 630 * SCALE
BLUE, NAVY, TEAL = "#4A90C4", "#1A2B4A", "#1E8A7B"
BODY, WHITE = "#5A6170", "#FFFFFF"

F = "/System/Library/Fonts/Supplemental/"
font_bold   = lambda s: ImageFont.truetype(F + "Arial Bold.ttf", s * SCALE)
font_italic = lambda s: ImageFont.truetype(F + "Arial Italic.ttf", s * SCALE)
font_reg    = lambda s: ImageFont.truetype(F + "Arial.ttf", s * SCALE)

def u(v):
    """design units -> output pixels"""
    return v * SCALE

root = os.path.join(os.path.dirname(__file__), "..")
img = Image.new("RGB", (W, H), WHITE)
d = ImageDraw.Draw(img)

# Left accent bar
d.rectangle([0, 0, u(14) - 1, H], fill=BLUE)

LEFT, TOP = u(90), u(64)

# Logo, scaled to 240px wide
logo = Image.open(os.path.join(root, "src/assets/sallysupport-logo.png")).convert("RGBA")
lw = u(240)
lh = round(logo.height * lw / logo.width)
img.paste(logo.resize((lw, lh), Image.LANCZOS), (LEFT, TOP), logo.resize((lw, lh), Image.LANCZOS))
y = TOP + lh + u(56)

def tracked(draw, xy, text, font, fill, tracking):
    """Pillow has no letter-spacing, so step glyph by glyph."""
    x, yy = xy
    for ch in text:
        draw.text((x, yy), ch, font=font, fill=fill)
        x += draw.textlength(ch, font=font) + tracking
    return x

def wrap(text, font, max_w):
    words, lines, cur = text.split(), [], ""
    for wd in words:
        trial = (cur + " " + wd).strip()
        if d.textlength(trial, font=font) <= max_w or not cur:
            cur = trial
        else:
            lines.append(cur); cur = wd
    if cur: lines.append(cur)
    return lines

# Eyebrow
f = font_bold(20)
tracked(d, (LEFT, y), "LIVE INDUSTRY BENCHMARK", f, BLUE, u(3))
y += u(24 + 18)

# Title
f = font_bold(64)
for line in wrap("The Admin Roadmap Report", f, u(760)):
    d.text((LEFT, y), line, font=f, fill=NAVY)
    y += u(67)
y += u(22)

# Byline
f = font_italic(26)
d.text((LEFT, y), "A report for home care agencies by SallySupport", font=f, fill=TEAL)
y += u(31 + 14)

# Body line
f = font_reg(21)
for line in wrap("How home care agencies hire, staff, and grow their office teams.", f, u(600)):
    d.text((LEFT, y), line, font=f, fill=BODY)
    y += u(32)

# Bar motif, bottom-right
bars = [(44, BLUE), (66, BLUE), (52, BLUE), (104, NAVY), (74, BLUE)]
bw, gap = u(30), u(10)
x = W - u(80) - (len(bars) * bw + (len(bars) - 1) * gap)
base = H - u(64)
for h, col in bars:
    d.rectangle([x, base - u(h), x + bw, base], fill=col)
    x += bw + gap

out = os.path.join(root, "public/og-image.png")
img.save(out)
print(f"Saved public/og-image.png ({img.width}x{img.height})")

File: scripts/generate-og-image.js
// Renders public/og-image.png (1200x630) through a real browser.
//
//   npm install --save-dev playwright
//   npx playwright install chromium
//   node scripts/generate-og-image.js
//
// Playwright is intentionally NOT in package.json: Vercel installs devDependencies
// during a build, and pulling a browser engine on every deploy to regenerate a
// static image nobody changed is a poor trade. Install it ad hoc when you need it.
//
// The machine this repo is usually maintained from has no Node, so the committed
// PNG was produced by generate-og-image.py, which draws the same layout with
// Pillow. If you change the design here, change it there too.
const fs = require('fs');
const path = require('path');
const { chromium } = require('playwright');

async function run() {
  // The logo lives in src/assets (it is imported by the app header), not public/.
  const logoPath = path.join(__dirname, '..', 'src', 'assets', 'sallysupport-logo.png');
  const logoBuffer = fs.readFileSync(logoPath);
  const logoDataUri = `data:image/png;base64,${logoBuffer.toString('base64')}`;

  const html = `
<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<style>
  * { margin:0; padding:0; box-sizing:border-box; }
  body {
    width: 1200px; height: 630px;
    background: #FFFFFF;
    font-family: -apple-system, "Segoe UI", Roboto, Helvetica, Arial, sans-serif;
    position: relative;
    overflow: hidden;
  }
  .bar { position: absolute; left:0; top:0; bottom:0; width:14px; background:#4A90C4; }
  .content { position: absolute; left:90px; top:64px; right:80px; }
  .logo { width:240px; height:auto; display:block; margin-bottom:56px; }
  .eyebrow {
    font-size:20px; font-weight:700; letter-spacing:0.15em; text-transform:uppercase;
    color:#4A90C4; margin-bottom:18px;
  }
  .title {
    font-size:64px; font-weight:800; color:#1A2B4A; margin-bottom:22px; line-height:1.05;
    max-width:760px; letter-spacing:-0.01em;
  }
  .byline { font-size:26px; font-style:italic; font-weight:500; color:#1E8A7B; margin-bottom:14px; }
  .body-line { font-size:21px; font-weight:400; color:#5A6170; max-width:600px; line-height:1.5; }
  .chart { position:absolute; right:80px; bottom:64px; display:flex; align-items:flex-end; gap:10px; }
  .chart .col { width:30px; background:#4A90C4; }
  .chart .col.tall { background:#1A2B4A; }
</style>
</head>
<body>
  <div class="bar"></div>
  <div class="content">
    <img class="logo" src="${logoDataUri}" />
    <p class="eyebrow">Live Industry Benchmark</p>
    <h1 class="title">The Admin Roadmap Report</h1>
    <p class="byline">A report for home care agencies by SallySupport</p>
    <p class="body-line">How home care agencies hire, staff, and grow their office teams.</p>
  </div>
  <div class="chart">
    <div class="col" style="height:44px"></div>
    <div class="col" style="height:66px"></div>
    <div class="col" style="height:52px"></div>
    <div class="col tall" style="height:104px"></div>
    <div class="col" style="height:74px"></div>
  </div>
</body>
</html>`;

  const browser = await chromium.launch();
  const page = await browser.newPage({ viewport: { width: 1200, height: 630 }, deviceScaleFactor: 1 });
  await page.setContent(html, { waitUntil: 'networkidle' });
  await page.screenshot({
    path: path.join(__dirname, '..', 'public', 'og-image.png'),
    clip: { x: 0, y: 0, width: 1200, height: 630 },
  });
  await browser.close();
  console.log('Saved public/og-image.png');
}

run().catch((err) => {
  console.error(err);
  process.exit(1);
});
